File: bwl.Imaging/GeometryProcessing/VectorTools.vb
Imports System.Drawing

Public Module VectorTools
    Public Class Vector
        Public Property X As Double
        Public Property Y As Double

        Public Sub New(X As Double, Y As Double)
            Me.X = X
            Me.Y = Y
        End Sub

        ''' <summary>
        ''' При вычислении вектора между (1,1) и (1,-1) этот метод дает - Pi/2! Это угловой переход от первого вектора ко второму
        ''' </summary>
        Public Shared Function AngleAB(A As Vector, B As Vector) As Double
            Return Math.Atan2(Cross(A, B), Dot(A, B))
        End Function

        Public ReadOnly Property Magnitude() As Double
            Get
                Return Math.Sqrt(Dot(Me, Me))
            End Get
        End Property

        Public Shared Function Dot(A As Vector, B As Vector) As Double
            Return A.X * B.X + A.Y * B.Y
        End Function

        Public Shared Function Cross(A As Vector, B As Vector) As Double
            Return A.X * B.Y - A.Y * B.X
        End Function
    End Class

    ''' <summary>
    ''' Тест: полигон пустой?
    ''' </summary>
    Public Function PolygonIsEmpty(polygon As PointF()) As Boolean
        If polygon.Length > 3 Then
            Dim firstPoint = polygon(0)
            For i = 1 To polygon.Length - 1
                Dim currentPoint = polygon(i)
                If currentPoint <> firstPoint Then
                    Return False
                End If
            Next
        End If
        Return True
    End Function

    ''' <summary>
    ''' Тест: точка находится внутри полигона?
    ''' </summary>
    Public Function PointInPolygon(point As PointF, polygon As PointF()) As Boolean
        Dim result = False
        Dim N = polygon.Length
        Dim j = N - 1
        For i = 0 To N - 1
            If ((polygon(i).Y <= point.Y) AndAlso (point.Y < polygon(j).Y)) OrElse ((polygon(j).Y <= point.Y) AndAlso (point.Y < polygon(i).Y)) Then
                If point.X < (polygon(j).X - polygon(i).X) * (point.Y - polygon(i).Y) / (polygon(j).Y - polygon(i).Y) + polygon(i).X Then
                    result = Not result
                End If
            End If
            j = i
        Next
        Return result
    End Function

    ''' <summary>
    ''' Нормализация длины вектора
    ''' </summary>    
    Public Function NormalizeVector(vector As PointF, Optional targetLen As Single = 1.0) As PointF
        Dim z = Math.Sqrt((vector.X * vector.X) + (vector.Y * vector.Y))
        Dim zNorm = targetLen / z 'Нормализатор к заданной длине
        Dim xn = vector.X * zNorm
        Dim yn = vector.Y * zNorm
        Return New PointF(CSng(xn), CSng(yn))
    End Function

    ''' <summary>
    ''' Проверка на нахождение вектора между векторами-границами
    ''' </summary>
    Public Function VectorInBounds(vector As PointF, vectorBound1 As PointF, vectorBound2 As PointF, Optional tol As Double = 0.017) As Boolean
        Dim boundsAngle = AngleBetweenVectors(vectorBound1, vectorBound2)
        Dim angle1 = AngleBetweenVectors(vector, vectorBound1)
        Dim angle2 = AngleBetweenVectors(vector, vectorBound2)
        Dim angle = angle1 + angle2
        Dim angleDiff = Math.Abs(boundsAngle - angle)
        Return angleDiff <= tol
    End Function

    ''' <summary>
    ''' Вычисление модуля угла между векторами №1
    ''' </summary>
    Public Function AngleBetweenVectors(vector1 As PointF, vector2 As PointF) As Double
        Return Math.Abs(AnglePathBetweenVectors(vector1, vector2))
    End Function

    ''' <summary>
    ''' Вычисление углового перехода между двумя векторами
    ''' </summary>    
    Public Function AnglePathBetweenVectors(vector1 As PointF, vector2 As PointF) As Double
        Dim v1 = New Vector(vector1.X, vector1.Y)
        Dim v2 = New Vector(vector2.X, vector2.Y)
        Return Vector.AngleAB(v1, v2)
    End Function
End Module

File: Bwl.Imaging.Skia/GeometryProcessing/VectorTools.cs
using SkiaSharp;
using System;


namespace Bwl.Imaging.Skia
{

    public static class VectorTools
    {
        public class Vector
        {
            public double X { get; set; }
            public double Y { get; set; }

            public Vector(double X, double Y)
            {
                this.X = X;
                this.Y = Y;
            }

            /// <summary>
            /// При вычислении вектора между (1,1) и (1,-1) этот метод дает - Pi/2! Это угловой переход от первого вектора ко второму
            /// </summary>
            public static double AngleAB(Vector A, Vector B)
            {
                return Math.Atan2(Cross(A, B), Dot(A, B));
            }

            public double Magnitude
            {
                get
                {
                    return Math.Sqrt(Dot(this, this));
                }
            }

            public static double Dot(Vector A, Vector B)
            {
                return A.X * B.X + A.Y * B.Y;
            }

            public static double Cross(Vector A, Vector B)
            {
                return A.X * B.Y - A.Y * B.X;
            }
        }

        /// <summary>
        /// Тест: полигон пустой?
        /// </summary>
        public static bool PolygonIsEmpty(SKPoint[] polygon)
        {
            if (polygon.Length > 3)
            {
                var firstPoint = polygon[0];
                for (int i = 1, loopTo = polygon.Length - 1; i <= loopTo; i++)
                {
                    var currentPoint = polygon[i];
                    if (currentPoint != firstPoint)
                    {
                        return false;
                    }
                }
            }
            return true;
        }

        /// <summary>
        /// Тест: точка находится внутри полигона?
        /// </summary>
        public static bool PointInPolygon(SKPoint point, SKPoint[] polygon)
        {
            bool result = false;
            int N = polygon.Length;
            int j = N - 1;
            for (int i = 0, loopTo = N - 1; i <= loopTo; i++)
            {
                if (polygon[i].Y <= point.Y && point.Y < polygon[j].Y || polygon[j].Y <= point.Y && point.Y < polygon[i].Y)
                {
                    if (point.X < (polygon[j].X - polygon[i].X) * (point.Y - polygon[i].Y) / (polygon[j].Y - polygon[i].Y) + polygon[i].X)
                    {
                        result = !result;
                    }
                }
                j = i;
            }
            return result;
        }

        /// <summary>
        /// Нормализация длины вектора
        /// </summary>
        public static SKPoint NormalizeVector(SKPoint vector, float targetLen = 1.0f)
        {
            double z = Math.Sqrt((double)(vector.X * vector.X + vector.Y * vector.Y));
            double zNorm = (double)targetLen / z; // Нормализатор к заданной длине
            double xn = (double)vector.X * zNorm;
            double yn = (double)vector.Y * zNorm;
            return new SKPoint((float)xn, (float)yn);
        }

        /// <summary>
        /// Проверка на нахождение вектора между векторами-границами
        /// </summary>
        public static bool VectorInBounds(SKPoint vector, SKPoint vectorBound1, SKPoint vectorBound2, double tol = 0.017d)
        {
            double boundsAngle = AngleBetweenVectors(vectorBound1, vectorBound2);
            double angle1 = AngleBetweenVectors(vector, vectorBound1);
            double angle2 = AngleBetweenVectors(vector, vectorBound2);
            double angle = angle1 + angle2;
            double angleDiff = Math.Abs(boundsAngle - angle);
            return angleDiff <= tol;
        }

        /// <summary>
        /// Вычисление модуля угла между векторами №1
        /// </summary>
        public static double AngleBetweenVectors(SKPoint vector1, SKPoint vector2)
        {
            return Math.Abs(AnglePathBetweenVectors(vector1, vector2));
        }

        /// <summary>
        /// Вычисление углового перехода между двумя векторами
        /// </summary>
        public static double AnglePathBetweenVectors(SKPoint vector1, SKPoint vector2)
        {
            var v1 = new Vector((double)vector1.X, (double)vector1.Y);
            var v2 = new Vector((double)vector2.X, (double)vector2.Y);
            return Vector.AngleAB(v1, v2);
        }
    }
}